src/models/llama/decoder.rs

```diff
@@ -0,0 +1,279 @@
+use std::sync::OnceLock;
+
+use regex::Regex;
+use serde::{Deserialize, Serialize};
+
+use crate::error::BoxedError;
+use crate::layers::activation::Activation;
+use crate::layers::attention::{AttentionHeads, QkvMode, SelfAttentionConfig};
+use crate::layers::embeddings::QueryKeyRotaryEmbeddingsConfig;
+use crate::layers::feedforward::PointwiseFeedForwardConfig;
+use crate::layers::layer_norm::RMSNormConfig;
+use crate::layers::transformer::{TransformerEmbeddingsConfig, TransformerLayerConfig};
+use crate::models::hf::FromHF;
+use crate::models::transformer::{TransformerDecoder, TransformerDecoderConfig};
+
+/// Llama decoder (Touvron et al., 2023).
+///
+/// See:
+/// * [LLaMA: Open and Efficient Foundation Language Models](https://arxiv.org/abs/2302.13971)
+/// * [Llama 2: Open Foundation and Fine-Tuned Chat Models](https://arxiv.org/abs/2307.09288)
+pub struct LlamaDecoder;
+
+/// HF Llama model types
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum HfModelType {
```

Comment:
Am I correct in my understanding that we are using an enum for the model type to facilitate a future `HFConfigWithModelType` trait  (and/or for deserialization convenience)?

Reply:
No, I wanted to avoid that every HF config struct also has to implement a trait, which increases implementation burden. The idea of this enum is to reject model configurations that do not have the right type.

Once we have auto models, they can have their own struct that deserializes just the model type.